File: 0289-Game of Life/0289-Game of Life.cpp
class Solution {
public:
    void gameOfLife(vector<vector<int>>& board) {
        int m = board.size();
        if (m == 0) {
            return;
        }
        int n = board[0].size();
        
        vector<int> deltaX = {-1, -1, -1, 0, 0, 1, 1, 1};
        vector<int> deltaY = {-1, 0, 1, -1, 1, -1, 0, 1};
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                int count = 0;
                for (int k = 0; k < deltaX.size(); ++k) {
                    int ni = i + deltaX[k];
                    int nj = j + deltaY[k];
                    if (ni >= 0 && ni < m && nj >= 0 && nj < n) {
                        if (board[ni][nj] & 1) {
                            ++count;
                        }
                    }
                }
                
                if (board[i][j] & 1) {
                    if (count == 2 || count == 3) {
                        board[i][j] |= 2;
                    }
                }
                else if (count == 3) {
                    board[i][j] |= 2;
                }
            }
        }
        
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                board[i][j] >>= 1;
            }
        }
    }
};

File: 0201-0300/0289-Game of Life/0289-Game of Life.py
class Solution:
    def gameOfLife(self, board: List[List[int]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        for i in range(len(board)):
            for j in range(len(board[0])):
                count = 0
                for r, c in ((i-1, j-1), (i-1, j), (i-1, j+1), (i, j-1), (i, j+1), (i+1, j-1), (i+1, j), (i+1, j+1)):
                    if 0 <= r < len(board) and 0 <= c < len(board[0]) and (board[r][c]&1):
                        count += 1
                        
                if board[i][j]&1:
                    if count == 2 or count == 3:
                        board[i][j] |= 2
                elif count == 3:
                    board[i][j] |= 2
                    
        for i in range(len(board)):
            for j in range(len(board[0])):
                board[i][j] >>= 1
